src/ui/public/directives/match_pairs.js

```diff
@@ -0,0 +1,101 @@
+import { uiModules } from 'ui/modules';
+const module = uiModules.get('kibana');
+
+/**
+ * This directively automatically handles matching pairs.
+ * Specifically, it does the following:
+ *
+ * 1. If the key is a closer, and the character in front of the cursor is the
+ *    same, simply move the cursor forward.
+ * 2. If the key is an opener, insert the opener at the beginning of the
+ *    selection, and the closer at the end of the selection, and move the
+ *    selection forward.
+ * 3. If the backspace is hit, and the characters before and after the cursor
+ *    are a pair, remove both characters and move the cursor backward.
+ */
+
+const pairs = ['()', '[]', '{}', `''`, '""'];
+const openers = pairs.map(pair => pair[0]);
+const closers = pairs.map(pair => pair[1]);
+
+module.directive('matchPairs', () => ({
+  restrict: 'A',
+  require: 'ngModel',
+  link: function (scope, elem, attrs, ngModel) {
+    elem.on('keydown', (e) => {
+      const { target, key, metaKey } = e;
+      const { value, selectionStart, selectionEnd } = target;
+
+      if (shouldMoveCursorForward()) {
+        moveCursorForward();
+      } else if (shouldInsertMatchingCloser()) {
+        insertMatchingCloser();
+      } else if (shouldRemovePair()) {
+        removePair();
+      }
+
+      function shouldMoveCursorForward() {
```

Comment:
All these functions should be hoisted out of the event handler, and you should just pass the key in. Defining them (and garbage collecting them) every time there's a key press is excessive.

Reply:
This is pretty simple for the `should...` functions but it seems odd for the others since they need `ngModel`. I pushed a commit just now... Let me know what you think: 57159424e2

Reply:
Yeah, that commit looks good. 👍